Mathlib/Algebra/Module/LinearMap.lean

```diff
@@ -0,0 +1,1252 @@
+
+export SemilinearMapClass (map_smulₛₗ)
+
+attribute [simp] map_smulₛₗ
+
+/-- `linear_map_class F R M M₂` asserts `F` is a type of bundled `R`-linear maps `M → M₂`.
+
+This is an abbreviation for `semilinear_map_class F (ring_hom.id R) M M₂`.
+-/
+abbrev LinearMapClass (F : Type _) (R M M₂ : outParam (Type _)) [Semiring R] [AddCommMonoid M]
+    [AddCommMonoid M₂] [Module R M] [Module R M₂] :=
+  SemilinearMapClass F (RingHom.id R) M M₂
+#align linear_map_class LinearMapClass
+
+namespace SemilinearMapClass
+
+variable (F : Type _)
+variable {_ : Semiring R} {_ : Semiring S}
+variable {_ : AddCommMonoid M} {_ : AddCommMonoid M₁} {_ : AddCommMonoid M₂} {_ : AddCommMonoid M₃}
+variable {_ : AddCommMonoid N₁} {_ : AddCommMonoid N₂} {_ : AddCommMonoid N₃}
+variable {_ : Module R M} {_ : Module R M₂} {_ : Module S M₃}
+variable {σ : R →+* S}
+
+-- Porting note: the `dangerousInstance` linter has become smarter about `outParam`s
+-- @[nolint dangerousInstance] -- `σ` is an `outParam` so it's not dangerous
+instance (priority := 100) addMonoidHomClass [SemilinearMapClass F σ M M₃] :
+    AddMonoidHomClass F M M₃ :=
+  { SemilinearMapClass.toAddHomClass with
+    coe := fun f ↦ (f : M → M₃)
+    map_zero := fun f ↦
+      show f 0 = 0 by
+        rw [← zero_smul R (0 : M), map_smulₛₗ]
+        simp }
+
+-- The `Semiring` should be an instance parameter but depends on outParams.
+-- If Lean 4 gets better support for instance params depending on outParams,
+-- we should be able to remove this nolint.
+@[nolint dangerousInstance]
+instance (priority := 100) distribMulActionHomClass [LinearMapClass F R M M₂] :
+    DistribMulActionHomClass F R M M₂ :=
+  { SemilinearMapClass.addMonoidHomClass F with
+    coe := fun f ↦ (f : M → M₂)
+    map_smul := fun f c x ↦ by rw [map_smulₛₗ, RingHom.id_apply] }
+
+variable {F} (f : F) [i : SemilinearMapClass F σ M M₃]
+
+theorem map_smul_inv {σ' : S →+* R} [RingHomInvPair σ σ'] (c : S) (x : M) :
+    c • f x = f (σ' c • x) := by simp
+#align semilinear_map_class.map_smul_inv SemilinearMapClass.map_smul_inv
+
+end SemilinearMapClass
+
+namespace LinearMap
+
+section AddCommMonoid
+
+variable [Semiring R] [Semiring S]
+
+section
+
+variable [AddCommMonoid M] [AddCommMonoid M₁] [AddCommMonoid M₂] [AddCommMonoid M₃]
+
+variable [AddCommMonoid N₁] [AddCommMonoid N₂] [AddCommMonoid N₃]
+
+variable [Module R M] [Module R M₂] [Module S M₃]
+
+variable {σ : R →+* S}
+
+instance : SemilinearMapClass (M →ₛₗ[σ] M₃) σ M M₃
+    where
+  coe f := f.toFun
+  coe_injective' f g h := by
+    cases f
+    cases g
+    congr
+    apply FunLike.coe_injective'
+    exact h
+  map_add f := f.map_add'
+  map_smulₛₗ := LinearMap.map_smul'
+
+-- Porting note: we don't port specialized `CoeFun` instances if there is `FunLike` instead
+#noalign linear_map.has_coe_to_fun
+
+-- Porting note: adding this instance prevents a timeout in `ext_ring_op`
+instance {σ : R →+* S} : FunLike (M →ₛₗ[σ] M₃) M (λ _ ↦ M₃) :=
+  { AddHomClass.toFunLike with }
```

Comment:
This instance could be removed if we can speed up `ext_ring_op` in some other way.

Reply:
(Indeed, with Gabriel's changes to Lean this can be deleted again.)